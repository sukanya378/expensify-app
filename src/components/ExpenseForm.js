import React from "react";
import { connect } from "react-redux";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            description: props.inititalExpenses ? props.inititalExpenses.description : "",
            amount: props.inititalExpenses ? (props.inititalExpenses.amount / 100).toString() : 0,
            note: props.inititalExpenses ? props.inititalExpenses.note : "",
            createdAt: props.inititalExpenses ? moment(props.inititalExpenses.createdAt) : moment(),
            calenderFocused: false,
            error: ""
        }
    }



    handleInputChange = (e) => {
        const { name, value } = e.target
        if (name === "amount") {
            if (!value || value.match(/^\d{1,}(\.\d{0,2})?$/)) this.setState({ [name]: value })
        }
        else this.setState({ [name]: value })
    }

    handledateChange = (createdAt) => {
        if (createdAt) this.setState({ createdAt })
    }

    onSubmit = (e) => {
        const { description, amount, note, createdAt } = this.state
        e.preventDefault();
        if (!description || !amount) this.setState({ error: "Please provide description and amount" })
        else {
            this.setState({ error: "" })
            this.props.onSubmit({
                description,
                amount: parseFloat(amount) * 100,
                note,
                createdAt: createdAt.valueOf()
            })
        }
    }

    handleFocusChange = ({ focused }) => {
        this.setState({ calenderFocused: focused })
    }


    render() {

        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <div>
                        <input
                            type="text"
                            placeholder="Description"
                            name="description"
                            autoFocus
                            onChange={this.handleInputChange}
                            value={this.state.description}
                        />

                        <div>
                            <input
                                type="text"
                                name="amount"
                                value={this.state.amount}
                                onChange={this.handleInputChange}
                            />
                        </div>

                        <textarea
                            placeholder="Add Note For Your expense"
                            name="note"
                            onChange={this.handleInputChange}
                        />

                        <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.handledateChange}
                            focused={this.state.calenderFocused}
                            onFocusChange={this.handleFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={(day) => false}
                        />



                    </div>
                    <button>Add Expense</button>
                    {this.state.error && <p>{this.state.error}</p>}
                </form>
            </div>
        )
    }
}

export default (ExpenseForm)