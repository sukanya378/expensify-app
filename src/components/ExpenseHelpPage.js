import React from "react"


const ExpenseHelpPage = () => (
    <div>This is my help page</div>
)

export default ExpenseHelpPage;