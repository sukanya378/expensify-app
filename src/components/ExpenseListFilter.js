import React, { useState } from "react";
import { connect } from "react-redux";
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from "../actions/filtersActions";
import { DateRangePicker } from "react-dates";


export const ExpenseListFilter = (props) => {
    const [calendarFocused, setCalendarFocused] = useState(null)
    const handleOnTextChange = (e) => {
        props.setTextFilter(e.target.value)
    }

    const handleOnSortChange = (e) => {
        (e.target.value === "date") ? props.sortByDate() : props.sortByAmount()
    }
    const onDatesChange = ({ startDate, endDate }) => {
        props.setStartDate(startDate)
        props.setEndDate(endDate)
    }
    return (
        <div>
            <input type="text" value={props.filters.text} onChange={handleOnTextChange}></input>
            <label>Sort By</label>
            <select onChange={handleOnSortChange} value={props.filters.sortBy}>
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>

            <DateRangePicker
                startDate={props.filters.startDate}
                endDate={props.filters.endDate}
                onDatesChange={onDatesChange}
                focusedInput={calendarFocused}
                onFocusChange={(focused) => setCalendarFocused(focused)}
                isOutsideRange={() => false}
                showClearDates={true}
                numberOfMonths={1}
            />
        </div>
    )
}



const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter : (text) => dispatch(setTextFilter(text)),
        sortByDate : () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate : (startDate) => dispatch(setStartDate(startDate)),
        setEndDate : (endDate) => dispatch(setEndDate(endDate))
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);