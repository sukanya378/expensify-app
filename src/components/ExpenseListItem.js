import React from "react";
import { NavLink } from "react-router-dom";

const ExpenseListItem = (props) => {
    const { id, description, amount, createdAt } = props.expense
    
    return (
        <div className="expense-list-item">
            <NavLink to={`/edit/${id}`} activeClassName="is-active"><h3>{description}</h3></NavLink>
            <div>{amount}</div>
            <div>{createdAt}</div>
            
        </div>
    )
}

export default ExpenseListItem;