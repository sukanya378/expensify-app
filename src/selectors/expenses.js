import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    expenses =  expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch;
    })

    expenses.sort((a,b) => {
        if(sortBy === "amount") return b.amount - a.amount
        else return b.createdAt - a.createdAt
    })
    return expenses
}

export default getVisibleExpenses