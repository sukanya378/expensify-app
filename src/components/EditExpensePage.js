import React from "react"
import ExpenseForm from "./expenseForm"
import { connect } from "react-redux"
import { editExpense } from "../actions/expensesActions"
import { removeExpense } from '../actions/expensesActions'

export class EditExpensePage extends React.Component {
    handleRemoveExpense = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                This is my edit page of id : {this.props.expense.id}
                <h1> Edit Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                    inititalExpenses={this.props.expense}
                />

                <button onClick={this.handleRemoveExpense}>Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.filter((expense) => expense.id === props.match.params.id)[0]
    }
}

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense(id))
})


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);