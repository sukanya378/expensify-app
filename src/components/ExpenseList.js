import React from "react";
import ExpenseListItem from "./ExpenseListItem";
import { connect } from "react-redux";
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
    return (
        <div>
            <h1>Expense List</h1>
            {props.expenses.length===0 ? <p>No Expense</p> : props.expenses.map((expense , itemIndex) => {
                return <ExpenseListItem expense = {expense} key={itemIndex}/>
            })}
            
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)     // Higher order component;