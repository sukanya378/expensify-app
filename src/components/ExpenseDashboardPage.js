import React from "react"
import ExpenseList from "./ExpenseList"
import ExpenseListFilter from "./ExpenseListFilter"


const ExpenseDashBoardPage = (props) => (
    <div>
        <ExpenseListFilter/>
        <ExpenseList expenses = {props.expenses}/>
    </div>
)
 
export default ExpenseDashBoardPage
