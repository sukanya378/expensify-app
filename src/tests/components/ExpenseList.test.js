import { ExpenseList } from "../../components/ExpenseList";
import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses"

test('should render Expense list with Expense', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render Expense list with no Expense', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})