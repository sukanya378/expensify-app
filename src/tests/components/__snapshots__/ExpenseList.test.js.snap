// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense list with Expense 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    expense={
      Object {
        "amount": 9000,
        "createdAt": 1686126588493,
        "description": "rent",
        "id": "1",
      }
    }
    key="0"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 200,
        "createdAt": 1686226488493,
        "description": "rent",
        "id": "2",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 10000,
        "createdAt": 1586226588493,
        "description": "Coffee",
        "id": "3",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 800,
        "createdAt": 1626226588493,
        "description": "Water Bill",
        "id": "4",
      }
    }
    key="3"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 1500,
        "createdAt": 1686226588403,
        "description": "Gas Bill",
        "id": "5",
      }
    }
    key="4"
  />
</div>
`;

exports[`should render Expense list with no Expense 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    expense={
      Object {
        "amount": 9000,
        "createdAt": 1686126588493,
        "description": "rent",
        "id": "1",
      }
    }
    key="0"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 200,
        "createdAt": 1686226488493,
        "description": "rent",
        "id": "2",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 10000,
        "createdAt": 1586226588493,
        "description": "Coffee",
        "id": "3",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 800,
        "createdAt": 1626226588493,
        "description": "Water Bill",
        "id": "4",
      }
    }
    key="3"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 1500,
        "createdAt": 1686226588403,
        "description": "Gas Bill",
        "id": "5",
      }
    }
    key="4"
  />
</div>
`;
